express-notes-api/server.ts

```diff
@@ -0,0 +1,120 @@
+import express from 'express';
+import { readFile, writeFile } from 'node:fs/promises';
+
+const app = express();
+
+type Entry = {
+  id: number;
+  content: string;
+};
+
+type Data = {
+  nextId: number;
+  notes: {
+    [key: number]: Entry;
+  };
+};
+
+async function write(obj: Data) {
+  const createData = JSON.stringify(obj, null, 2);
+  await writeFile('./data.json', `${createData} \n`);
+}
+
+app.use(express.json());
+
+app.get('/api/notes', async (req, res) => {
+  try {
+    const jsonData = await readFile('./data.json', 'utf-8');
+    const dataParsed: Data = JSON.parse(jsonData);
+    const entryArray: Entry[] = [];
+    for (const entry in dataParsed.notes) {
+      entryArray.push(dataParsed.notes[entry]);
+    }
+    res.json(entryArray);
+  } catch (error) {
+    console.error(error);
+    res.status(500).json({ error: 'an unexpected error occurred' });
+  }
+});
+
+app.get('/api/notes/:id', async (req, res) => {
+  try {
+    const jsonData = await readFile('./data.json', 'utf-8');
+    const dataParsed: Data = JSON.parse(jsonData);
+    const id = Number(req.params.id);
+    if (id <= 0 || isNaN(id)) {
+      res.status(400).json({ Error: 'ID must be a positive integer!' });
+    } else if (!dataParsed.notes[id]) {
+      res.status(404).json({ Error: `No entry exists with ID: ${id}` });
+    } else {
+      res.status(200).json(dataParsed.notes[id]);
+    }
+  } catch (error) {
+    console.error(error);
+    res.status(500).json({ error: 'an unexpected error occurred' });
+  }
+});
+
+app.post('/api/notes', async (req, res) => {
+  const jsonData = await readFile('./data.json', 'utf-8');
+  const dataParsed: Data = JSON.parse(jsonData);
+  try {
+    if (!req.body.content) {
+      res.status(400).json({ Error: 'Content is a required field' });
+    }
+    dataParsed.notes[dataParsed.nextId] = req.body;
+    dataParsed.notes[dataParsed.nextId].id = dataParsed.nextId;
+    write(dataParsed);
+    res.status(201).json(dataParsed.notes[dataParsed.nextId]);
+    dataParsed.nextId++;
+  } catch (error) {
+    console.error(error);
+    res.status(500).json({ Error: 'Unexpected error' });
+  }
+});
+
+app.delete('/api/notes/:id', async (req, res) => {
+  const jsonData = await readFile('./data.json', 'utf-8');
+  const dataParsed: Data = JSON.parse(jsonData);
+  const id = Number(req.params.id);
+  try {
```

Comment:
You should put the try around `readFile` since you want to catch any errors that throws. Remember that any time you use `await` you must wrap it in try/catch.

Looks like all the handlers have this problem.